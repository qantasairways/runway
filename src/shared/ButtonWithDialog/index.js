import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Transition } from 'react-transition-group';

import noop from '../../utils/noop';
import { layout, mq, fontFamily } from '../../theme/airways';

import Dialog from './components/Dialog';
import Button, { ButtonContent } from './components/Button';
import MediaQueryDetector from '../../components/MediaQueryDetector';

const getRootContainerStyles = hasDialogDimensions => ({
  ...(hasDialogDimensions && {
    position: 'relative'
  })
});

function addQComZIndex(zIndex) {
  const main = document.querySelector('main');
  if (main) {
    main.style.zIndex = zIndex;
  }
}

export const transitionStylesSlideUp = {
  entering: {
    top: '100%'
  },
  entered: {
    top: 0
  },
  exiting: {
    top: '100%'
  }
};

export const dialogStylesFullScreen = {
  label: 'runway__dialog',
  fontFamily: fontFamily.main,
  background: 'white',
  boxSizing: 'border-box',
  height: '100%',
  left: 0,
  position: 'fixed',
  overflow: 'auto',
  top: '100%',
  transition: `top 300ms ease-in-out`,
  width: '100%',
  zIndex: 1000,
  display: 'flex',
  flexDirection: 'column'
};

class ButtonWithDialog extends Component {
  state = {
    open: false
  };

  onEntered = () => {
    this.props.onOpen();
    addQComZIndex('initial');

    if (this.props.closeOnBlur) {
      document.addEventListener('click', this.handleClickOutside);
    }

    setTimeout(() => {
      if (this.focusElement) {
        this.focusElement.focus();
      }
    });
  };

  onExited = () => {
    this.props.onClose();
    addQComZIndex(null);

    if (this.props.closeOnBlur) {
      document.removeEventListener('click', this.handleClickOutside);
    }

    if (this.fieldButton) {
      this.fieldButton.focus();
    }
  };

  handleClickOutside = event => {
    if (this.wrapper && !this.wrapper.contains(event.target)) {
      this.closeDialog();
    }
  };

  openDialog = () => {
    this.setState({
      open: true
    });
  };

  closeDialog = () => {
    this.setState({
      open: false
    });
  };

  setFocusElementRef = el => {
    this.focusElement = el;
  };

  setButtonRef = el => {
    this.fieldButton = el;
  };

  renderHeader = () =>
    this.props.renderHeader({
      closeDialog: this.closeDialog,
      setFocusElementRef: this.focusElementRef
    });

  renderFooter = () =>
    this.props.renderFooter({
      closeDialog: this.closeDialog,
      setFocusElementRef: this.focusElementRef
    });

  renderButton = () => {
    const { open } = this.state;
    const {
      children,
      dialogAriaLabel,
      renderHeader,
      ...buttonProps
    } = this.props;

    return (
      <Button
        onClick={this.openDialog}
        setButtonRef={this.setButtonRef}
        open={open}
        {...buttonProps}
      />
    );
  };

  collectRef = el => {
    this.wrapper = el;
  };

  render() {
    const {
      children,
      dialogAriaLabel,
      dialogStyles,
      transitionStyles,
      contentPadding,
      hasDialogDimensions
    } = this.props;

    const { open } = this.state;

    const content =
      typeof children === 'function'
        ? children({
            closeDialog: this.closeDialog,
            setFocusElementRef: this.setFocusElementRef
          })
        : children;

    return (
      <div
        css={getRootContainerStyles({ hasDialogDimensions })}
        ref={this.collectRef}
      >
        {this.renderButton()}
        <MediaQueryDetector query={mq.medium}>
          {atLeastTablet => {
            const isFullScreen = !hasDialogDimensions || !atLeastTablet;
            return (
              <Transition
                in={open}
                enter={isFullScreen}
                exit={isFullScreen}
                onExited={this.onExited}
                onEntered={this.onEntered}
                timeout={300}
                unmountOnExit
                mountOnEnter
              >
                {transitionState => (
                  <Dialog
                    lockBgScroll={isFullScreen}
                    renderHeader={isFullScreen ? this.renderHeader : noop}
                    renderFooter={this.renderFooter}
                    contentPadding={contentPadding}
                    transitionStyles={transitionStyles}
                    dialogStyles={dialogStyles}
                    dialogAriaLabel={dialogAriaLabel}
                    content={content}
                    transitionState={transitionState}
                  />
                )}
              </Transition>
            );
          }}
        </MediaQueryDetector>
      </div>
    );
  }
}

ButtonWithDialog.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  renderButtonValue: PropTypes.func.isRequired,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  onBlur: PropTypes.func,
  closeOnBlur: PropTypes.bool,
  closeAriaLabel: PropTypes.string,
  dialogAriaLabel: PropTypes.string,
  renderHeader: PropTypes.func,
  renderFooter: PropTypes.func,
  dialogStyles: PropTypes.shape().isRequired,
  transitionStyles: PropTypes.shape({
    entering: PropTypes.shape.isRequired,
    entered: PropTypes.shape.isRequired,
    exiting: PropTypes.shape.isRequired
  }).isRequired,
  contentPadding: PropTypes.string,
  hasDialogDimensions: PropTypes.bool
};

ButtonWithDialog.defaultProps = {
  children: null,
  onClose: noop,
  onOpen: noop,
  onBlur: noop,
  closeAriaLabel: '',
  closeOnBlur: true,
  renderHeader: noop,
  renderFooter: noop,
  dialogAriaLabel: '',
  contentPadding: `0 ${layout.gutter}`,
  hasDialogDimensions: false
};

export { ButtonContent };

export default ButtonWithDialog;
