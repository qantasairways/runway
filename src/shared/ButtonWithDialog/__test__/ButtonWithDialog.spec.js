import React from 'react';
import { mount, shallow } from 'enzyme';
import ButtonWithDialog from '..';

function openDialog(component) {
  component.instance().setState({ open: true });
}

describe('ButtonWithDialog', () => {
  let component;

  it('renders correctly with defaults', () => {
    component = shallow(
      <ButtonWithDialog
        renderButtonValue={() => <div>button content</div>}
        dialogStyles={{}}
        transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props', () => {
    beforeAll(() => {
      component = shallow(
        <ButtonWithDialog
          closeAriaLabel="Close Aria Label"
          dialogAriaLabel="Dialog Aria Label"
          onClose={e => e}
          onBlur={e => e}
          placeHolder="placeholder"
          renderButtonValue={() => <div>button value</div>}
          renderHeader={() => <div>header</div>}
          dialogStyles={{}}
          transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
          contentPadding="1px"
        >
          {'children'}
        </ButtonWithDialog>
      );
      openDialog(component);
    });

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with one child element', () => {
    component = shallow(
      <ButtonWithDialog
        renderButtonValue={() => <div>button content</div>}
        dialogStyles={{}}
        transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
      >
        <span>Child Element</span>
      </ButtonWithDialog>
    );
    openDialog(component);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with one child function', () => {
    component = mount(
      <ButtonWithDialog
        renderButtonValue={() => <div>button content</div>}
        dialogStyles={{}}
        transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
      >
        {({ closeDialog }) => <button type="button" onClick={closeDialog} />}
      </ButtonWithDialog>
    );
    openDialog(component);

    expect(component).toMatchSnapshot();
  });

  describe('onEntered()', () => {
    beforeEach(() => {
      document.addEventListener = jest.fn();
      component = mount(
        <ButtonWithDialog
          dialogAriaLabel="Dialog Aria Label"
          onClose={e => e}
          onBlur={e => e}
          placeHolder="placeholder"
          renderButtonValue={() => <div>button value</div>}
          renderHeader={() => <div>header</div>}
          dialogStyles={{}}
          transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
          contentPadding="1px"
        >
          {({ setFocusElementRef }) => (
            <div>
              <button type="button" ref={setFocusElementRef} />
            </div>
          )}
        </ButtonWithDialog>
      );
      openDialog(component);
      jest.spyOn(component.instance().focusElement, 'focus');
      component.instance().onEntered();
      jest.useFakeTimers();
    });

    afterEach(() => {
      document.addEventListener.mockReset();
      component.instance().focusElement.focus.mockReset();
    });

    it('adds click event listener', () => {
      const addEventListenerCalls = document.addEventListener.mock.calls;

      expect(addEventListenerCalls.length).toBe(1);
      expect(addEventListenerCalls[0][0]).toBe('click');
      expect(addEventListenerCalls[0][1]).toBe(
        component.instance().handleClickOutside
      );
    });

    it('focuses the element with ref this.focusElement', () => {
      jest.runAllTimers();
      expect(component.instance().focusElement.focus).toHaveBeenCalled();
    });
  });

  describe('onExited()', () => {
    const mockOnClose = jest.fn();

    beforeEach(() => {
      document.removeEventListener = jest.fn();
      component = mount(
        <ButtonWithDialog
          closeAriaLabel="Close Aria Label"
          dialogAriaLabel="Dialog Aria Label"
          onClose={mockOnClose}
          onBlur={e => e}
          placeHolder="placeholder"
          renderButtonValue={() => <div>button value</div>}
          renderHeader={() => <div>header</div>}
          dialogStyles={{}}
          transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
          contentPadding="1px"
        />
      );
      jest.spyOn(component.instance().fieldButton, 'focus');
      component.instance().onExited();
    });

    afterEach(() => {
      document.removeEventListener.mockReset();
      mockOnClose.mockReset();
      component.instance().fieldButton.focus.mockReset();
    });

    it('removes click event listener', () => {
      const removeEventListenerCalls = document.removeEventListener.mock.calls;

      expect(removeEventListenerCalls.length).toBe(1);
      expect(removeEventListenerCalls[0][0]).toBe('click');
      expect(removeEventListenerCalls[0][1]).toBe(
        component.instance().handleClickOutside
      );
    });

    it('calls this.props.onClose()', () => {
      expect(mockOnClose.mock.calls.length).toBe(1);
    });

    it('focuses the field button', () => {
      expect(component.instance().fieldButton.focus).toHaveBeenCalled();
    });
  });

  describe('handleClickOutside()', () => {
    beforeEach(() => {
      component = mount(
        <ButtonWithDialog
          closeAriaLabel="Close Aria Label"
          dialogAriaLabel="Dialog Aria Label"
          onClose={e => e}
          onBlur={e => e}
          placeHolder="placeholder"
          renderButtonValue={() => <div>button value</div>}
          renderHeader={() => <div>header</div>}
          dialogStyles={{}}
          transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
          contentPadding="1px"
        />
      );
      openDialog(component);
      jest.spyOn(component.instance(), 'closeDialog');
    });

    afterEach(() => {
      component.instance().closeDialog.mockReset();
    });

    it('calls closeDialog() if the click was outside', () => {
      const componentInstance = component.instance();

      componentInstance.handleClickOutside({ target: document });
      expect(componentInstance.closeDialog).toHaveBeenCalled();
    });

    it('does nothing if the click was inside', () => {
      const componentInstance = component.instance();

      componentInstance.handleClickOutside({
        target: componentInstance.wrapper
      });
      expect(componentInstance.closeDialog).not.toHaveBeenCalled();
    });
  });

  describe('openDialog()', () => {
    beforeEach(() => {
      component = mount(
        <ButtonWithDialog
          closeAriaLabel="Close Aria Label"
          dialogAriaLabel="Dialog Aria Label"
          onClose={e => e}
          onBlur={e => e}
          placeHolder="placeholder"
          renderButtonValue={() => <div>button value</div>}
          renderHeader={() => <div>header</div>}
          dialogStyles={{}}
          transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
          contentPadding="1px"
        />
      );
    });

    it('opens the popup', () => {
      component.instance().openDialog();

      expect(component.state('open')).toBe(true);
    });
  });

  describe('closeDialog()', () => {
    beforeEach(() => {
      component = mount(
        <ButtonWithDialog
          closeAriaLabel="Close Aria Label"
          dialogAriaLabel="Dialog Aria Label"
          onClose={e => e}
          onBlur={e => e}
          placeHolder="placeholder"
          renderButtonValue={() => <div>button value</div>}
          renderHeader={() => <div>header</div>}
          dialogStyles={{}}
          transitionStyles={{ entering: {}, entered: {}, exiting: {} }}
          contentPadding="1px"
        />
      );
    });

    it('closes the popup', () => {
      const componentInstance = component.instance();

      componentInstance.setState({ open: true });
      componentInstance.closeDialog();

      expect(component.state('open')).toBe(false);
    });
  });
});
