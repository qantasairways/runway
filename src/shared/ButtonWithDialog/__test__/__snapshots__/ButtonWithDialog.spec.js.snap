// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonWithDialog renders correctly with defaults 1`] = `
.emotion-0 {
  position: relative;
}

<div
  className="emotion-0"
>
  <Button
    Icon={null}
    buttonLabel=""
    closeAriaLabel=""
    closeOnBlur={true}
    dialogStyles={Object {}}
    hasDialogDimensions={false}
    iconLabelButtonValue={null}
    onBeforeClose={[Function]}
    onBlur={[Function]}
    onClick={[Function]}
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    placeHolder=""
    renderButtonValue={[Function]}
    renderFooter={[Function]}
    setButtonRef={[Function]}
    shouldAddScrollLockClass={false}
    transitionStyles={
      Object {
        "entered": Object {},
        "entering": Object {},
        "exiting": Object {},
      }
    }
  />
  <MediaQueryDetector
    query="@media only screen and (min-width: 768px)"
  >
    <Component />
  </MediaQueryDetector>
</div>
`;

exports[`ButtonWithDialog renders correctly with one child element 1`] = `
.emotion-0 {
  position: relative;
}

<div
  className="emotion-0"
>
  <Button
    Icon={null}
    buttonLabel=""
    closeAriaLabel=""
    closeOnBlur={true}
    dialogStyles={Object {}}
    hasDialogDimensions={false}
    iconLabelButtonValue={null}
    onBeforeClose={[Function]}
    onBlur={[Function]}
    onClick={[Function]}
    onClose={[Function]}
    onOpen={[Function]}
    open={true}
    placeHolder=""
    renderButtonValue={[Function]}
    renderFooter={[Function]}
    setButtonRef={[Function]}
    shouldAddScrollLockClass={false}
    transitionStyles={
      Object {
        "entered": Object {},
        "entering": Object {},
        "exiting": Object {},
      }
    }
  />
  <MediaQueryDetector
    query="@media only screen and (min-width: 768px)"
  >
    <Component />
  </MediaQueryDetector>
</div>
`;

exports[`ButtonWithDialog renders correctly with one child function 1`] = `
.emotion-3 {
  position: relative;
}

.emotion-2 {
  position: relative;
  height: 6rem;
  width: 100%;
  max-width: 100%;
  border: 0;
  padding: 0 15px;
  cursor: pointer;
  background-color: #323232;
  color: #fff;
  font-family: "Ciutadella",sans-serif;
  font-size: 1.125rem;
  font-weight: 400;
}

.emotion-1 {
  top: 5px;
  left: 7px;
  color: white;
  text-transform: none;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  position: absolute;
  top: 5px;
  left: 7px;
  color: white;
  font-size: 1rem;
}

<ButtonWithDialog
  closeAriaLabel=""
  closeOnBlur={true}
  dialogAriaLabel=""
  dialogStyles={Object {}}
  hasDialogDimensions={false}
  onBeforeClose={[Function]}
  onBlur={[Function]}
  onClose={[Function]}
  onOpen={[Function]}
  renderButtonValue={[Function]}
  renderFooter={[Function]}
  renderHeader={[Function]}
  shouldAddScrollLockClass={false}
  transitionStyles={
    Object {
      "entered": Object {},
      "entering": Object {},
      "exiting": Object {},
    }
  }
>
  <div
    className="emotion-3"
  >
    <Button
      Icon={null}
      buttonLabel=""
      closeAriaLabel=""
      closeOnBlur={true}
      dialogStyles={Object {}}
      hasDialogDimensions={false}
      iconLabelButtonValue={null}
      onBeforeClose={[Function]}
      onBlur={[Function]}
      onClick={[Function]}
      onClose={[Function]}
      onOpen={[Function]}
      open={false}
      placeHolder=""
      renderButtonValue={[Function]}
      renderFooter={[Function]}
      setButtonRef={[Function]}
      shouldAddScrollLockClass={false}
      transitionStyles={
        Object {
          "entered": Object {},
          "entering": Object {},
          "exiting": Object {},
        }
      }
    >
      <button
        aria-expanded={false}
        aria-haspopup="dialog"
        className="emotion-2"
        onBlur={[Function]}
        onClick={[Function]}
        type="button"
      >
        <div
          className="emotion-1"
        >
          <span
            className="emotion-0"
          />
          <div>
            button content
          </div>
        </div>
      </button>
    </Button>
    <MediaQueryDetector
      query="@media only screen and (min-width: 768px)"
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={300}
        unmountOnExit={true}
      />
    </MediaQueryDetector>
  </div>
</ButtonWithDialog>
`;

exports[`ButtonWithDialog renders correctly with props 1`] = `
.emotion-0 {
  position: relative;
}

<div
  className="emotion-0"
>
  <Button
    Icon={null}
    buttonLabel=""
    closeAriaLabel=""
    closeOnBlur={true}
    dialogStyles={Object {}}
    hasDialogDimensions={false}
    iconLabelButtonValue={null}
    onBeforeClose={[Function]}
    onBlur={[Function]}
    onClick={[Function]}
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    placeHolder=""
    renderButtonValue={[Function]}
    renderFooter={[Function]}
    setButtonRef={[Function]}
    shouldAddScrollLockClass={false}
    transitionStyles={
      Object {
        "entered": Object {},
        "entering": Object {},
        "exiting": Object {},
      }
    }
  />
  <MediaQueryDetector
    query="@media only screen and (min-width: 768px)"
  >
    <Component />
  </MediaQueryDetector>
</div>
`;
